paperqa/core.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import annotations
+
+import json
+import re
+from collections.abc import Callable
+from typing import Any
+
+from .llms import Chain
+from .types import Context, LLMResult, Text
+from .utils import get_score, strip_citations
+
+
+def llm_parse_json(text: str) -> dict:
+    """Read LLM output and extract JSON data from it."""
+    # fetch from markdown ```json if present
+    text = text.strip().split("```json")[-1].split("```")[0]
+    # split anything before the first {
+    text = "{" + text.split("{", 1)[-1]
+    # split anything after the last }
+    text = text.rsplit("}", 1)[0] + "}"
+
+    # escape new lines within strings
+    def replace_newlines(match: re.Match) -> str:
+        return match.group(0).replace("\n", "\\n")
+
+    # Match anything between double quotes
+    # including escaped quotes and other escaped characters.
+    # https://regex101.com/r/VFcDmB/1
+    pattern = r'"(?:[^"\\]|\\.)*"'
+    text = re.sub(pattern, replace_newlines, text)
+
+    return json.loads(text)
+
+
+async def map_fxn_summary(
+    text: Text,
+    question: str,
+    chain: Chain | None,
```

Comment:
Wdyt of removing the `Chain` class? Imo we're using a term from LangChain that is really just a prompt with some pre or post processing

Maybe I am wrong here, but I feel like the term "chain" falls in the same bucket as "text-to-action" -- not really that useful of a term, more of a buzzword

Reply:
Discussed in person - no better name for closure for executing LLM + prompt + config